coalib/core/PersistentHash.py

```diff
@@ -1,7 +1,53 @@
+from collections import Iterable
 from hashlib import sha1
 import pickle
 
 
+def tuple_replace(tup, x, y):
+    tup_list = list(tup)
+    for element in tup_list:
+        if element == x:
+            tup_list[tup_list.index(element)] = y
+    new_tuple = tuple(tup_list)
+    return new_tuple
+
+
+def preserve_order(obj):
+    args_tuple = obj[0]
+    sorted_args = args_tuple
+
+    # iterates over the tuple of positional arguments
+    # looks for sets and dicts, sort them and put them
+    # back into the tuple.
+    for element in sorted_args:
+        if isinstance(element, Iterable):
+            if any(element) and isinstance(element, dict):
+                sorted_args = tuple_replace(sorted_args, element,
+                                            dict(sorted(element.items())))
```

Comment:
Do you have to replace items actually? I mean just create a new tuple / list and further operate on that.

This has maybe some flaws, e.g. a tuple with `((1, 3),)` would have the same persistent hash as the dict `{1: 3}`, but maybe that's not so important for us? Or we could somehow modify the internal data so we can distinguish different object types. Anyway, not so important, but I wouldn't replace things inside instances.